### PR TITLE
Remove panics in 'to protobuf' logic of the converter

### LINKED ISSUE #50: Remove panic from server code
Currently, panic is called when an exception such as an error occurs in the server code. The server goes down, but it is easier to grasp the error situation.

However, in the production environment, it should be changed to return an error instead of panic so that the server does not stop.

### PR DESCRIPTION
**Pull request by dc7303**
**What this PR does / why we need it**:
If an error occurs in the converter, the converter must return an error
and allow the caller to decide how to handle it.

**Which issue(s) this PR fixes**:

Fixes #50 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation**:

```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

### CODE REVIEW

File: api/converter/to_bytes.go [hunk 1]
```diff
@@ -26,15 +28,20 @@ import (
 
 // ObjectToBytes converts the given object to byte array.
 func ObjectToBytes(obj *json.Object) ([]byte, error) {
-	bytes, err := proto.Marshal(toJSONElement(obj))
+	elem, err := toJSONElement(obj)
```
Comment (hackerwins) [inline api/converter/to_bytes.go:31]: To distinguish protobuf message from other objects, the prefix `pbXXX` is used. How about changing this to `pbElem`?

```suggestion
	pbElem, err := toJSONElement(obj)
```

File: api/converter/to_pb.go [hunk 2]
```diff
@@ -52,17 +59,23 @@ func toCheckpoint(cp *checkpoint.Checkpoint) *api.Checkpoint {
 	}
 }
 
-func toChanges(changes []*change.Change) []*api.Change {
+func toChanges(changes []*change.Change) ([]*api.Change, error) {
 	var pbChanges []*api.Change
+
 	for _, c := range changes {
+		operations, err := ToOperations(c.Operations())
```
Comment (hackerwins) [inline api/converter/to_pb.go:66]:
```suggestion
		pbOperations, err := ToOperations(c.Operations())
```

File: api/converter/to_bytes.go [hunk 3]
```diff
@@ -43,53 +50,65 @@ func toJSONElement(elem json.Element) *api.JSONElement {
...
-
-	// It occurs when the implementation is omitted in the converter after
-	// adding a new data type during development. So we explicitly fire the
-	// panic.
-	panic("fail to encode JSONElement to protobuf")
 }
 
-func toJSONObject(obj *json.Object) *api.JSONElement {
+func toJSONObject(obj *json.Object) (*api.JSONElement, error) {
+	rhtNodes, err := toRHTNodes(obj.RHTNodes())
```
Comment (hackerwins) [inline api/converter/to_bytes.go:64]:
```suggestion
	pbRHTNodes, err := toRHTNodes(obj.RHTNodes())
```

File: api/converter/to_bytes.go [hunk 4]
```diff
@@ -43,53 +50,65 @@ func toJSONElement(elem json.Element) *api.JSONElement {
...
 			RemovedAt: toTimeTicket(obj.RemovedAt()),
 		}},
 	}
-	return pbElem
+	return pbElem, nil
 }
 
-func toJSONArray(arr *json.Array) *api.JSONElement {
+func toJSONArray(arr *json.Array) (*api.JSONElement, error) {
+	rgaNodes, err := toRGANodes(arr.RGANodes())
```
Comment (hackerwins) [inline api/converter/to_bytes.go:81]:
```suggestion
	pbRGANodes, err := toRGANodes(arr.RGANodes())
```

File: api/converter/to_bytes.go [hunk 5]
```diff
@@ -43,53 +50,65 @@ func toJSONElement(elem json.Element) *api.JSONElement {
...
 			RemovedAt: toTimeTicket(arr.RemovedAt()),
 		}},
 	}
-	return pbElem
+	return pbElem, nil
 }
 
-func toPrimitive(primitive *json.Primitive) *api.JSONElement {
+func toPrimitive(primitive *json.Primitive) (*api.JSONElement, error) {
+	valueType, err := toValueType(primitive.ValueType())
```
Comment (hackerwins) [inline api/converter/to_bytes.go:98]:
```suggestion
	pbValueType, err := toValueType(primitive.ValueType())
```

File: api/converter/to_bytes.go [hunk 6]
```diff
@@ -114,37 +133,52 @@ func toRichText(text *json.RichText) *api.JSONElement {
 	}
 }
 
-func toCounter(counter *json.Counter) *api.JSONElement {
+func toCounter(counter *json.Counter) (*api.JSONElement, error) {
+	counterType, err := toCounterType(counter.ValueType())
```
Comment (hackerwins) [inline api/converter/to_bytes.go:137]:
```suggestion
	pbCounterType, err := toCounterType(counter.ValueType())
```

File: api/converter/to_bytes.go [hunk 7]
```diff
@@ -114,37 +133,52 @@ func toRichText(text *json.RichText) *api.JSONElement {
...
 		}},
-	}
+	}, nil
 }
 
-func toRHTNodes(rhtNodes []*json.RHTPQMapNode) []*api.RHTNode {
+func toRHTNodes(rhtNodes []*json.RHTPQMapNode) ([]*api.RHTNode, error) {
 	var pbRHTNodes []*api.RHTNode
 	for _, rhtNode := range rhtNodes {
+		elem, err := toJSONElement(rhtNode.Element())
```
Comment (hackerwins) [inline api/converter/to_bytes.go:156]:
```suggestion
		pbElem, err := toJSONElement(rhtNode.Element())
```

File: api/converter/to_bytes.go [hunk 8]
```diff
@@ -114,37 +133,52 @@ func toRichText(text *json.RichText) *api.JSONElement {
...
 	}
-	return pbRHTNodes
+	return pbRHTNodes, nil
 }
 
-func toRGANodes(rgaNodes []*json.RGATreeListNode) []*api.RGANode {
+func toRGANodes(rgaNodes []*json.RGATreeListNode) ([]*api.RGANode, error) {
 	var pbRGANodes []*api.RGANode
 	for _, rgaNode := range rgaNodes {
+		elem, err := toJSONElement(rgaNode.Element())
```
Comment (hackerwins) [inline api/converter/to_bytes.go:172]:
```suggestion
		pbElem, err := toJSONElement(rgaNode.Element())
```

File: api/converter/to_pb.go [hunk 9]
```diff
@@ -28,14 +30,19 @@ import (
 )
 
 // ToChangePack converts the given model format to Protobuf format.
-func ToChangePack(pack *change.Pack) *api.ChangePack {
+func ToChangePack(pack *change.Pack) (*api.ChangePack, error) {
+	changes, err := toChanges(pack.Changes)
```
Comment (hackerwins) [inline api/converter/to_pb.go:34]:
```suggestion
	pbChanges, err := toChanges(pack.Changes)
```

File: api/converter/to_pb.go [hunk 10]
```diff
@@ -99,156 +112,222 @@ func ToClientsMap(clientsMap map[string][]string) map[string]*api.Clients {
...
 		}
 		pbOperations = append(pbOperations, pbOperation)
 	}
 
-	return pbOperations
+	return pbOperations, nil
+}
+
+func toSet(set *operation.Set) (*api.Operation_Set_, error) {
+	elem, err := toJSONElementSimple(set.Value())
```
Comment (hackerwins) [inline api/converter/to_pb.go:167]:
```suggestion
	pbElem, err := toJSONElementSimple(set.Value())
```

File: api/converter/to_pb.go [hunk 11]
```diff
@@ -99,156 +112,222 @@ func ToClientsMap(clientsMap map[string][]string) map[string]*api.Clients {
...
+			ParentCreatedAt: toTimeTicket(set.ParentCreatedAt()),
+			Key:             set.Key(),
+			Value:           elem,
+			ExecutedAt:      toTimeTicket(set.ExecutedAt()),
+		},
+	}, nil
+}
+
+func toAdd(add *operation.Add) (*api.Operation_Add_, error) {
+	elem, err := toJSONElementSimple(add.Value())
```
Comment (hackerwins) [inline api/converter/to_pb.go:183]:
```suggestion
	pbElem, err := toJSONElementSimple(add.Value())
```

File: client/client.go [hunk 12]
```diff
@@ -186,9 +186,14 @@ func (c *Client) Attach(ctx context.Context, doc *document.Document) error {
 
 	doc.SetActor(c.id)
 
+	changePack, err := converter.ToChangePack(doc.CreateChangePack())
```
Comment (hackerwins) [inline client/client.go:189]:
```suggestion
	pbChangePack, err := converter.ToChangePack(doc.CreateChangePack())
```

File: api/converter/to_pb.go [hunk 13]
```diff
@@ -99,156 +112,222 @@ func ToClientsMap(clientsMap map[string][]string) map[string]*api.Clients {
...
+		return api.EventType_DOCUMENTS_CHANGED, nil
 	case types.DocumentsWatchedEvent:
-		return api.EventType_DOCUMENTS_WATCHED
+		return api.EventType_DOCUMENTS_WATCHED, nil
 	case types.DocumentsUnwatchedEvent:
-		return api.EventType_DOCUMENTS_UNWATCHED
+		return api.EventType_DOCUMENTS_UNWATCHED, nil
 	default:
-		panic("unsupported event")
+		return 0, fmt.Errorf("%v: %w", eventType, ErrUnsupportedEventType)
```
Comment (hackerwins) [inline api/converter/to_pb.go:124]: `EventType` is declared as a string. How about changing this line like below?

```suggestion
		return 0, fmt.Errorf("%s: %w", eventType, ErrUnsupportedEventType)
```

File: api/converter/to_bytes.go [hunk 14]
```diff
@@ -43,53 +50,65 @@ func toJSONElement(elem json.Element) *api.JSONElement {
...
 	case *json.Text:
-		return toText(elem)
+		return toText(elem), nil
 	case *json.RichText:
-		return toRichText(elem)
+		return toRichText(elem), nil
 	case *json.Counter:
 		return toCounter(elem)
+	default:
+		return nil, fmt.Errorf("%s: %w", elem, ErrUnsupportedElement)
```
Comment (hackerwins) [inline api/converter/to_bytes.go:59]: `elem` is an instance of struct `json.Element`, will this be properly formatted?

File: api/converter/to_pb.go [hunk 15]
```diff
@@ -99,156 +112,222 @@ func ToClientsMap(clientsMap map[string][]string) map[string]*api.Clients {
...
-			Type:      toCounterType(elem.ValueType()),
+			Type:      counterType,
 			CreatedAt: toTimeTicket(elem.CreatedAt()),
 			Value:     elem.Bytes(),
-		}
+		}, nil
 	}
-	panic("fail to encode JSONElement to protobuf")
+
+	return nil, fmt.Errorf("%d, %w", elem, ErrUnsupportedElement)
```
Comment (hackerwins) [inline api/converter/to_pb.go:330]: will this line be properly formatted?

**Review by hackerwins — CHANGES_REQUESTED** [review]
Thank you for your work. 👍 I left two types of comments.

- Whether `fmt.Errorf` is properly formatted.
- Apply `pbXXX` name to protobuf message.

File: api/converter/to_bytes.go (hunk 14 above)
Comment (dc7303) [reply api/converter/to_bytes.go:59]: Oh! my mistake. The log is very messy when output like this. How about modifying it like the code below?

```go
fmt.Errorf("%v: %w", reflect.TypeOf(elem), ErrUnsupportedElement)
```

Output example:
`*json.Primitive: unsupported element`

Comment (hackerwins) [reply api/converter/to_bytes.go:59]: 👍 The message got better.

**Comment by dc7303** [comment]
@hackerwins 
Thank you for the meticulous review! Corrected the issues you suggested :)

**Review by hackerwins — APPROVED** [review]
LGTM. 👍 Thanks for contributing.

**Pull request merged**